### PR TITLE
[single-implicit-asset-job] methods on `CodeLocation` to get job partitions, partition tags, and partition config, with asset selection

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

This builds on top of https://github.com/dagster-io/dagster/pull/23527 to do two related things:
- On the host process side, refactors `CodeLocation` methods and call sites to use job names and selected asset keys instead of partition set names.
- On the user code process side, updates the implementations of grpc APIs that fetch partition information to pass selected asset keys to the `JobDefinition`.

### CODE REVIEW

**Review by gibsondan — COMMENTED** [review]
sandy can you say more about what we are doing here to ensure that selected_asset_keys is only used/needed when there are newer versions of user code? how does that requirement get threaded through?

**Comment by sryza** [comment]
> sandy can you say more about what we are doing here to ensure that selected_asset_keys is only used/needed when there are newer versions of user code? how does that requirement get threaded through?

#### Why it's only _needed_ when there are newer versions of user code: 

1.8.1 will be the first Dagster version where it's possible for a job to have multiple `PartitionsDefinition`s "inside" it. Prior to this, every job either had a single `partitions_def` or had no partitioning at all.

In 1.8.1, it will be possible for an asset job to target assets with different `PartitionsDefinition`s. Even though those jobs don't have a single `PartitionsDefinition`, it's sensible to ask "what partitions can I launch a run of this job with if I'm only selecting asset1?"  This is what the new `selected_asset_keys` parameter allows us to do.

`selected_asset_keys` is not needed and can be safely ignored prior to 1.8.1 because the answer to "what partitions can I launch a run of this job with?" doesn't depend on what assets are selected, because the assets in the job will never have different `PartitionsDefinition`s.

#### Why it's only _used_ when there are newer versions of user code

This might reveal a naive misunderstanding of how things work, but I thought the answer to this was basically "only versions of user code on or 1.8.1 after will know about this `selected_asset_keys`, so only these versions can/will use it".

**Comment by gibsondan** [comment]
"selected_asset_keys is not needed and can be safely ignored prior to 1.8.1 " answers my question - this scheme works as long as the grpc client isn't relying on the selected_asset_keys to do filtering, even on older versions. (For example, if this were like the asset_selection argument which causes certain runs to subset the assets that are being materialized - it would be bad if that argument was sometimes dropped on the floor by the server and ignored, because then you would launch a run for every asset instead of just the assets being selected - but it sounds like this argument is not like that)

File: python_modules/dagster/dagster/_core/remote_representation/code_location.py
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/remote_representation/code_location.py:475]: This seems like an optimization that could have big implications if it was being used a lot and was removed (specifically, a lot more calls to user code). Although the fact that it is being called on the InProcessCodeLocation is confusing since you're sort of already in user code at that point (and InProcessCodeLocation isn't really used outside of tests and maybe the CLI), so maybe a moot point

File: python_modules/dagster/dagster/_core/remote_representation/code_location.py
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/remote_representation/code_location.py:883]: ah OK, we are still keeping the optimization mentioned above in the place where it might actually matter, cool

File: python_modules/dagster/dagster/_grpc/impl.py
Comment (gibsondan) [inline python_modules/dagster/dagster/_grpc/impl.py:514]: was removing this part intentional / related to this PR?

File: python_modules/dagster/dagster/_grpc/types.py
Comment (gibsondan) [inline python_modules/dagster/dagster/_grpc/types.py:420]: This will break when users upgrade their dagit/daemon versions without also upgrading their user code - we should make sure that we are prepared for user reports of that and maybe ensure that our docs are clear about exactly what our back-compat guarantees are in OSS - in the past we have been kind of wishy washy about it and I think we even have tests in our backcompat suite that exercise new versions of OSS with old versions of dagit

Maybe it's something we could detect in dagit and warn about so that the resulting error is less cryptic? Code servers tell us what dagster versions they are using

**Review by gibsondan — APPROVED** [review]
i'm not actually seeing anything set the new selected_asset_keys field other than derived it from the legacy partition set - is there still a change coming that has the frontend vary that field in some way that it wasn't before?

File: python_modules/dagster/dagster/_grpc/types.py
Comment (sryza) [reply python_modules/dagster/dagster/_grpc/types.py:420]: > This will break when users upgrade their dagit/daemon versions without also upgrading their user code

Do you mean it will break when users upgrade their user code without upgrading their dagit/daemon versions?

Comment (gibsondan) [reply python_modules/dagster/dagster/_grpc/types.py:420]: Yes sorry, I had that reversed

File: python_modules/dagster/dagster/_core/remote_representation/code_location.py
Comment (sryza) [reply python_modules/dagster/dagster/_core/remote_representation/code_location.py:475]: linking this here for future readers: https://github.com/dagster-io/dagster/pull/23491/files#r1715290754

File: python_modules/dagster/dagster/_grpc/types.py
Comment (sryza) [reply python_modules/dagster/dagster/_grpc/types.py]: > maybe ensure that our docs are clear about exactly what our back-compat guarantees are in OSS

Here's a PR that adds a section on this: https://github.com/dagster-io/dagster/pull/23734.

> Maybe it's something we could detect in dagit and warn about so that the resulting error is less cryptic?

Thoughts on where we would add this? The gRPC call that (re)loads the definitions? Do you think we should add this before merging this PR?

Comment (gibsondan) [reply python_modules/dagster/dagster/_grpc/types.py]: I don’t think it needs to block the PR - if there’s a known exception it will emit we could catch it and add “this error may be due to version skew” in the error message.

File: python_modules/dagster/dagster/_grpc/impl.py
Comment (sryza) [reply python_modules/dagster/dagster/_grpc/impl.py:514]: It was intentional, though I should have flagged it and could be easily nudged to break it out into its own PR.

If you look at `_instance_from_ref_for_dynamic_partitions`, there's this weirdness right now where we only construct an instance ref if the `PartitionsDefinition` is a `DynamicPartitionsDefinition`, because otherwise we want the user to not need a dagster.yaml in their code server.

HOWEVER, part of what's happening in this PR is that we abstract away the `PartitionsDefinition` from this code path (this allows a future where someone executes a run for "2024-04-23" for asset1 and asset2, but those assets have different daily `PartitionsDefinition`s with different start dates). So now we don't have the `PartitionsDefinition` to look at and see if it's dynamic.

We could add some sort of `JobDefinition.partitions_def_is_dynamic(selected_asset_keys)` method to retain the existing behavior. HOWEVER HOWEVER, the reason I deleted this code is that I question whether this validation is really important. It result in an extra round trip to the DB. And this weirdness with the instance ref. And the backend code paths that invoke it already know the partition exists.

**Comment by sryza** [comment]
> i'm not actually seeing anything set the new selected_asset_keys field other than derived it from the legacy partition set - is there still a change coming that has the frontend vary that field in some way that it wasn't before?

**Comment by sryza** [comment]
> i'm not actually seeing anything set the new selected_asset_keys field other than derived it from the legacy partition set - is there still a change coming that has the frontend vary that field in some way that it wasn't before?

Oops yes, I should have linked here: https://github.com/dagster-io/dagster/pull/23494. And then @bengotow has a branch for the frontend changes.

**Pull request merged**
